genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/KillServiceImpl.java

```diff
@@ -33,54 +38,54 @@
+    KillServiceImpl(final ExecutionContext executionContext, final AgentProperties agentProperties) {
         this(
-            applicationEventPublisher,
-            agentProperties, DEFAULT_ACTION
+            executionContext,
+            agentProperties,
+            Thread::new
         );
     }
 
+    @VisibleForTesting
     KillServiceImpl(
-        final ApplicationEventPublisher applicationEventPublisher,
+        final ExecutionContext executionContext,
         final AgentProperties agentProperties,
-        final Runnable emergencyTerminationAction
+        final ThreadFactory threadFactory
     ) {
-        this.applicationEventPublisher = applicationEventPublisher;
+        this.executionContext = executionContext;
         this.agentProperties = agentProperties;
-        this.emergencyTerminationThread = new Thread(
-            () -> {
-                log.debug("Emergency shutdown countdown started");
-                try {
-                    Thread.sleep(this.agentProperties.getEmergencyShutdownDelay().toMillis());
-                } catch (InterruptedException e) {
-                    log.warn("Interrupted during delayed emergency countdown");
-                }
-                log.warn("Emergency shutdown now");
-                emergencyTerminationAction.run();
-            },
-            EMERGENCY_TERMINATION_THREAD_NAME
-        );
+        this.threadFactory = threadFactory;
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public void kill(final KillSource killSource) {
-        ConsoleLog.getLogger().info("Job kill requested (source: {})", killSource.name());
-        log.debug("Publishing kill event");
-        this.applicationEventPublisher.publishEvent(new KillEvent(killSource));
+        if (this.killed.compareAndSet(false, true)) {
+            ConsoleLog.getLogger().info("Job kill requested (source: {})", killSource.name());
+            this.executionContext.getStateMachine().kill(killSource);
+            this.threadFactory.newThread(this::emergencyStop).start();
+        }
+    }
 
-        // Start emergency termination thread, if not already running
-        if (!this.emergencyTerminationThread.isAlive()) {
-            this.emergencyTerminationThread.start();
+    /**
+     * This task is launched when a call to kill() is received.
+     * It's extra insurance that the JVM eventually will shut down.
+     * Hopefully, execution terminates due to regular shutdown procedures before this mechanism kicks-in.
+     */
+    @SuppressFBWarnings("DM_EXIT") // For calling System.exit
+    private void emergencyStop() {
+        try {
+            Thread.sleep(this.agentProperties.getEmergencyShutdownDelay().toMillis());
+        } catch (InterruptedException e) {
+            log.warn("Emergency shutdown thread interrupted");
         }
+        System.exit(ExitCode.EXEC_ABORTED.getCode());
```

Comment:
May be overcomplicating things here but there are [other ways to shut down a spring application](https://www.baeldung.com/spring-boot-shutdown#exit) that don't necessarily involve explicitly calling `System.exit`. Up to you whether it's worth it.

Reply:
This is really a last resort mechanism that should never kick in. 
If it ends up being used, it's because the Spring shutdown somehow got stuck.